feat: 독서 기록 V2 API 및 대분류/세부 감정 시스템 구현

move-hoon opened the pull request:
## 🔗 관련 이슈
- Close #145

## 📘 작업 유형
- [x] ✨ Feature (기능 추가)
- [ ] 🐞 Bugfix (버그 수정)
- [x] 🔧 Refactor (코드 리팩토링)
- [ ] ⚙️ Chore (환경 설정)
- [ ] 📝 Docs (문서 작성 및 수정)
- [ ] ✅ Test (기능 테스트)
- [ ] 🎨 style (코드 스타일 수정)

## 📙 작업 내역
### 감정 시스템 V2 구조
- 대분류 감정 5개: 따뜻함, 즐거움, 슬픔, 깨달음, 기타
- 세부 감정 태그: 대분류별 여러 세부 감정 지원 (다중 선택 가능)

### Domain Layer
- [PrimaryEmotion](cci:2://file:///Users/mgrv/Desktop/mgrv/Reed-Server/domain/src/main/kotlin/org/yapp/domain/readingrecord/PrimaryEmotion.kt:2:0-16:1) enum 추가 (대분류 감정)
- [DetailTag](cci:2://file:///Users/mgrv/Desktop/mgrv/Reed-Server/domain/src/main/kotlin/org/yapp/domain/detailtag/DetailTag.kt:7:0-57:1) 도메인 및 Repository 추가 (세부 감정)
- [ReadingRecordDetailTag](cci:2://file:///Users/mgrv/Desktop/mgrv/Reed-Server/infra/src/main/kotlin/org/yapp/infra/readingrecorddetailtag/entity/ReadingRecordDetailTagEntity.kt:13:0-72:1) 도메인 및 Repository 추가 (독서기록-세부감정 연결)
- [DetailTagDomainService](cci:2://file:///Users/mgrv/Desktop/mgrv/Reed-Server/domain/src/main/kotlin/org/yapp/domain/detailtag/DetailTagDomainService.kt:6:0-26:1), [ReadingRecordDetailTagDomainService](cci:2://file:///Users/mgrv/Desktop/mgrv/Reed-Server/domain/src/main/kotlin/org/yapp/domain/readingrecorddetailtag/ReadingRecordDetailTagDomainService.kt:5:0-32:1) 추가
- [ReadingRecord](cci:1://file:///Users/mgrv/Desktop/mgrv/Reed-Server/apis/src/main/kotlin/org/yapp/apis/readingrecord/controller/ReadingRecordControllerV2.kt:50:4-65:5)에 `primaryEmotion` 필드 추가, `pageNumber` nullable 변경

### Infra Layer
- DB 마이그레이션 스크립트 추가 (`V20251224_001`)
  - `detail_tags` 테이블 생성
  - `reading_record_detail_tags` 테이블 생성
  - `reading_records.primary_emotion` 컬럼 추가
- [DetailTagEntity](cci:2://file:///Users/mgrv/Desktop/mgrv/Reed-Server/infra/src/main/kotlin/org/yapp/infra/detailtag/entity/DetailTagEntity.kt:10:0-71:1), [ReadingRecordDetailTagEntity](cci:2://file:///Users/mgrv/Desktop/mgrv/Reed-Server/infra/src/main/kotlin/org/yapp/infra/readingrecorddetailtag/entity/ReadingRecordDetailTagEntity.kt:13:0-72:1) 추가
- Repository 구현체 추가

### APIs Layer (V2)
- `POST /api/v2/reading-records/{userBookId}` - 독서 기록 생성
- `GET /api/v2/reading-records/detail/{readingRecordId}` - 독서 기록 상세 조회
- `GET /api/v2/reading-records/{userBookId}` - 독서 기록 목록 조회
- `PUT /api/v2/reading-records/{readingRecordId}` - 독서 기록 수정
- `DELETE /api/v2/reading-records/{readingRecordId}` - 독서 기록 삭제
- `GET /api/v2/emotions` - 감정 목록 조회 API

### DTO 리팩토링
- V2 DTO에 private constructor + 정적 팩토리 메서드 패턴 일관성 있게 적용
- PrimaryEmotionDto, DetailEmotionDto, EmotionGroupDto, EmotionDetailDto를 nested class로 리팩토링
- CreateReadingRecordRequestV2에 @NotNull validation 추가

### primaryEmotion 변경 시 데이터 일관성 보장
- 대분류 감정 변경 시 세부 감정 태그 처리 로직 추가
- 대분류 변경 + 새 태그 제공 → 새 태그로 교체
- 대분류 변경 + 태그 미제공 → 기존 태그 삭제 (불일치 방지)
- 대분류 유지 + 새 태그 제공 → 새 태그로 교체
- 대분류 유지 + 태그 미제공 → 기존 태그 유지

### 소유권 검증 강화 (V1/V2 공통)
- getReadingRecordDetail, updateReadingRecord, deleteReadingRecord에 소유권 검증 추가
- USER_BOOK_ACCESS_DENIED (403 FORBIDDEN) 에러 코드 추가
- 다른 사용자의 독서 기록에 무단 접근 시 403 반환

### API 문서 개선
- 정렬 파라미터 우선순위 명확화 (sort 파라미터 vs Pageable.sort)
- pageNumber 필드에 "(선택)" 표시 추가

### V1 호환성
- 기존 V1 API는 그대로 유지
- V1 Response의 pageNumber nullable 변경
- V1에도 동일한 소유권 검증 적용

## 🧪 테스트 내역
- [x] 브라우저/기기에서 동작 확인
- [ ] 엣지 케이스 테스트 완료
- [x] 기존 기능 영향 없음

## ✅ PR 체크리스트
- [x] 커밋 메시지가 명확합니다
- [x] PR 제목이 컨벤션에 맞습니다
- [x] 관련 이슈 번호를 작성했습니다
- [x] 기능이 정상적으로 작동합니다
- [x] 불필요한 코드를 제거했습니다

## 💬 추가 설명 or 리뷰 포인트
- **커밋 구조**: 17개 커밋이 구현 논리 순서대로 작성되어 있어 코드 리뷰 시 흐름 파악이 용이합니다
- **DomainService 분리**: V2에서는 `ApplicationService`가 여러 [DomainService](cci:2://file:///Users/mgrv/Desktop/mgrv/Reed-Server/domain/src/main/kotlin/org/yapp/domain/detailtag/DetailTagDomainService.kt:6:0-26:1)를 조합하는 구조로 설계했습니다
- **TODO**: [ReadingRecordDomainService](cci:2://file:///Users/mgrv/Desktop/mgrv/Reed-Server/domain/src/main/kotlin/org/yapp/domain/readingrecord/ReadingRecordDomainService.kt:18:0-296:1)의 V1 관련 의존성은 추후 분리 예정 (주석 표시됨)
- **세부 감정 데이터**: 마이그레이션에서 초기 세부 감정 데이터가 INSERT됩니다
- **보안 강화**: V1/V2 모두 독서 기록 접근 시 소유권 검증 추가 (403 FORBIDDEN)

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * 감정 목록 조회 API 추가
  * 독서기록 V2 API 도입(주감정·세부감정 태그 지원, CRUD, 페이징/정렬)

* **Enhancements**
  * 요청/응답 포맷 확장: 주감정, 세부감정 포함 및 pageNumber 선택형 처리
  * 세부감정(DetailTag) 도입 및 연결 테이블·관리 로직 추가
  * 마이그레이션으로 기존 데이터에 주감정 칼럼 및 초기 세부감정 데이터 추가

* **Bug Fixes / Changes**
  * 사용자 책 권한 오류 코드 및 메시지 변경 (접근 거부)

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

move-hoon committed:
[BOOK-480] refactor: apis - emotionTags는 빈 리스트도 허용이 되기 때문에 valid 패턴 제거

move-hoon committed:
[BOOK-480] feat: infra - 감정 시스템 DB 스키마 마이그레이션 추가

move-hoon committed:
[BOOK-480] feat: domain - PrimaryEmotion 대분류 감정 enum 추가

move-hoon committed:
[BOOK-480] feat: domain - DetailTag 세부감정 도메인 및 Repository 인터페이스 추가

move-hoon committed:
[BOOK-480] feat: domain - ReadingRecordDetailTag 연결 도메인 및 Repository 인터페이스 추가

move-hoon committed:
[BOOK-480] feat: infra - DetailTag 엔티티 및 Repository 구현체 추가

move-hoon committed:
[BOOK-480] feat: infra - ReadingRecordDetailTag 엔티티 및 Repository 구현체 추가

move-hoon committed:
[BOOK-480] feat: domain, infra - ReadingRecord에 primaryEmotion 필드 추가

move-hoon committed:
[BOOK-480] feat: domain - ReadingRecordInfoVO에 V2용 detailEmotions 필드 추가

move-hoon committed:
[BOOK-480] feat: domain - DetailTagDomainService 추가

move-hoon committed:
[BOOK-480] feat: domain - ReadingRecordDetailTagDomainService 추가

move-hoon committed:
[BOOK-480] refactor: domain - ReadingRecordDomainService V2 메서드 추가 및 의존성 분리

move-hoon committed:
[BOOK-480] feat: apis - V2 Request/Response DTO 추가

move-hoon committed:
[BOOK-480] feat: apis - ReadingRecordServiceV2 ApplicationService 추가

move-hoon committed:
[BOOK-480] feat: apis - ReadingRecordUseCaseV2 유스케이스 추가

move-hoon committed:
[BOOK-480] feat: apis - ReadingRecordControllerV2 컨트롤러 추가

move-hoon committed:
[BOOK-480] feat: apis - EmotionController 감정 목록 조회 API 추가

move-hoon committed:
[BOOK-480] fix: apis - V1 ReadingRecordResponse pageNumber nullable 호환 수정

minwoo1999 commented on apis/src/main/kotlin/org/yapp/apis/readingrecord/service/ReadingRecordServiceV2.kt:
```diff
@@ -0,0 +1,180 @@
+        userId: UUID,
+        readingRecordId: UUID,
+        request: UpdateReadingRecordRequestV2
+    ): ReadingRecordResponseV2 {
+        val existingRecord = readingRecordDomainService.findById(readingRecordId)
+        val newPrimaryEmotion = request.primaryEmotion ?: existingRecord.primaryEmotion
+
+        // Validate detail emotion tags if provided
+        if (!request.detailEmotionTagIds.isNullOrEmpty()) {
+            validateDetailEmotionTags(request.detailEmotionTagIds, newPrimaryEmotion)
+        }
```
이부분 한번 확인 부탁드립니다!

move-hoon replied:
@minwoo1999 해결했습니다!

minwoo1999 commented on apis/src/main/kotlin/org/yapp/apis/emotion/dto/response/EmotionListResponse.kt:
```diff
@@ -0,0 +1,52 @@
+import java.util.UUID
+
+@Schema(name = "EmotionListResponse", description = "감정 목록 응답")
+data class EmotionListResponse(
```
private construct 안붙어있는 dto클래스들이 몇몇개 보여서 확인 부탁드릴게요 !!

move-hoon replied:
추가해두었습니다!

minwoo1999 reviewed (approved):
고생하셨습니다 :)

move-hoon committed:
[BOOK-480] refactor: apis - private constructor 및 정적 팩토리 메서드로 패턴 변경

move-hoon committed:
[BOOK-480] fix: apis - 독서 기록 목록 조회 (V2) 명세 오류 수정

move-hoon committed:
[BOOK-480] fix: apis - 누락된 NotNull 어노테이션 추가

move-hoon committed:
[BOOK-480] chore: apis - pageNumber가 선택 사항임을 명시

move-hoon committed:
[BOOK-480] fix: apis - primaryEmotion 변경 시 detailEmotion 데이터 일관성 보장

변경 사항:
- primaryEmotion 변경 + 새 detailTags 제공 → 새 태그로 교체
- primaryEmotion 변경 + detailTags 미제공 → 기존 태그 삭제 (불일치 방지)
- primaryEmotion 유지 + 새 detailTags 제공 → 새 태그로 교체
- primaryEmotion 유지 + detailTags 미제공 → 기존 태그 유지
리팩토링:
- updateDetailEmotionTags() private 메서드로 분리
- isPrimaryEmotionChanged() 헬퍼 메서드 추가

move-hoon committed:
[BOOK-480] refactor: apis - 독시 기록에 대한 소유권 검증 로직 추가

move-hoon committed:
[BOOK-480] refactor: apis - readingRecords를 불변 리스트로 변경

move-hoon committed:
[BOOK-480] chore: apis - 403 에러 명세 추가

move-hoon committed:
[BOOK-480] refactor: apis - 기존 단언 패턴으로 valid 메서드 구현방식 변경

move-hoon committed:
[BOOK-480] chore: apis - Void 타입 대신 Kotlin의 Unit 타입으로 변경

move-hoon committed:
[BOOK-480] refactor: apis - 코드레빗 리뷰 반영
